File: src/readability.js
export function checkReadability(markdown, maxGrade = 7) {
  const grade = fleschKincaidGrade(markdown);
  return {
    grade,
    maxGrade,
    ok: grade <= maxGrade
  };
}

export function fleschKincaidGrade(markdown) {
  const text = normalize(markdown);
  const sentences = Math.max(1, (text.match(/[.!?]+(?:\s|$)/g) ?? []).length);
  const words = text.match(/[A-Za-z]+(?:'[A-Za-z]+)?/g) ?? [];
  if (words.length === 0) return 0;

  const syllables = words.reduce((count, word) => count + countSyllables(word), 0);
  const grade = 0.39 * (words.length / sentences) + 11.8 * (syllables / words.length) - 15.59;
  return Math.max(0, Number(grade.toFixed(2)));
}

function normalize(markdown) {
  return markdown
    .replace(/```[\s\S]*?```/g, " ")
    .replace(/`[^`]*`/g, " ")
    .replace(/https?:\/\/\S+/g, " ")
    .replace(/[#>*_[\]()-]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function countSyllables(word) {
  const clean = word.toLowerCase().replace(/[^a-z]/g, "");
  if (!clean) return 0;
  if (clean.length <= 3) return 1;

  const withoutSilentE = clean.replace(/e\b/, "");
  const groups = withoutSilentE.match(/[aeiouy]+/g);
  return Math.max(1, groups ? groups.length : 1);
}

File: src/readability.rs
#[derive(Clone, Debug, PartialEq)]
pub struct Readability {
    pub grade: f64,
    pub max_grade: f64,
    pub ok: bool,
}

pub fn check_readability(markdown: &str) -> Readability {
    let max_grade = 7.0;
    let grade = flesch_kincaid_grade(markdown);
    Readability {
        grade,
        max_grade,
        ok: grade <= max_grade,
    }
}

pub fn flesch_kincaid_grade(markdown: &str) -> f64 {
    let text = normalize(markdown);
    let sentences = text
        .chars()
        .filter(|char| matches!(char, '.' | '!' | '?'))
        .count()
        .max(1) as f64;
    let words = words(&text);
    if words.is_empty() {
        return 0.0;
    }
    let syllables: usize = words.iter().map(|word| count_syllables(word)).sum();
    let grade = 0.39 * (words.len() as f64 / sentences)
        + 11.8 * (syllables as f64 / words.len() as f64)
        - 15.59;
    round2(grade.max(0.0))
}

fn normalize(markdown: &str) -> String {
    let mut output = String::new();
    let mut in_code = false;

    for line in markdown.lines() {
        if line.trim_start().starts_with("```") {
            in_code = !in_code;
            continue;
        }
        if in_code {
            continue;
        }
        output.push_str(line);
        output.push(' ');
    }

    output
        .split_whitespace()
        .filter(|part| !part.starts_with("http://") && !part.starts_with("https://"))
        .collect::<Vec<_>>()
        .join(" ")
        .replace(['#', '>', '*', '_', '[', ']', '(', ')', '-', '`'], " ")
}

fn words(text: &str) -> Vec<String> {
    text.split(|char: char| !char.is_ascii_alphabetic() && char != '\'')
        .filter(|part| part.chars().any(|char| char.is_ascii_alphabetic()))
        .map(ToString::to_string)
        .collect()
}

fn count_syllables(word: &str) -> usize {
    let clean = word
        .to_ascii_lowercase()
        .chars()
        .filter(|char| char.is_ascii_alphabetic())
        .collect::<String>();
    if clean.is_empty() {
        return 0;
    }
    if clean.len() <= 3 {
        return 1;
    }

    let clean = clean.strip_suffix('e').unwrap_or(&clean);
    let mut count = 0;
    let mut last_was_vowel = false;
    for char in clean.chars() {
        let is_vowel = matches!(char, 'a' | 'e' | 'i' | 'o' | 'u' | 'y');
        if is_vowel && !last_was_vowel {
            count += 1;
        }
        last_was_vowel = is_vowel;
    }
    count.max(1)
}

fn round2(value: f64) -> f64 {
    (value * 100.0).round() / 100.0
}
